File: competence-database/src/main/scala/uzuzjmd/competence/owl/dao/SelfAssessment.scala
package uzuzjmd.competence.owl.dao

import uzuzjmd.competence.owl.access.CompOntologyManager
import uzuzjmd.competence.owl.ontology.CompOntClass
import uzuzjmd.competence.owl.access.CompOntologyAccessScala
import uzuzjmd.competence.owl.ontology.CompObjectProperties

case class SelfAssessment(comp: CompOntologyManager, competence: Competence, user: User, val assmentIndex: java.lang.Integer = null) extends CompetenceOntologyDao(comp, CompOntClass.SelfAssessment, CompOntologyAccessScala.createIdentifierForAssessment(user, competence)) {
  def ASSESSMENTINDEX = "assessmentIndex" // 1-4 in liferay implementation
  def LEARNINGGOAL = "learningGoal" // boolean

  @Override
  def getFullDao(): SelfAssessment = {
    return new SelfAssessment(comp, competence, user, getDataFieldInt(ASSESSMENTINDEX))
  }

  @Override
  def persistMore() {
    if (assmentIndex != null) {
      addDataField(ASSESSMENTINDEX, assmentIndex)
    }
  }

  @Override
  def deleteMore() {

  }

  def updateIndex(assmentIndex: Integer) {
    addDataField(ASSESSMENTINDEX, assmentIndex)
  }

  def addCompetenceToAssessment(competence: Competence) {
    createEdgeWith(CompObjectProperties.AssessmentOfCompetence, competence)
  }

  def addUserToAssessment(user: User) {
    createEdgeWith(CompObjectProperties.AssessmentOfUser, user)
  }

  def getAssmentIndexInProgress(): Integer = {
    if (getDataField(ASSESSMENTINDEX) == null) {
      return 0
    }
    return new Integer(getDataFieldInt(ASSESSMENTINDEX) * 25)
  }

  def getAssmentIndex(): Integer = {
    if (getDataField(ASSESSMENTINDEX) == null) {
      return 0
    }
    return new Integer(getDataFieldInt(ASSESSMENTINDEX))
  }

  def getLearningGoal(): Boolean = {
    if (getDataField(LEARNINGGOAL) == null) {
      return false;
    }
    return getDataFieldBoolean(LEARNINGGOAL)
  }

}

File: competence-database/src/main/scala/uzuzjmd/competence/monopersistence/daos/SelfAssessment.java
package uzuzjmd.competence.monopersistence.daos;

import uzuzjmd.competence.monopersistence.Cascadable;
import uzuzjmd.competence.persistence.ontology.CompObjectProperties;

/**
 * Created by dehne on 11.01.2016.
 */
public class SelfAssessment extends AbstractSelfAssessment implements Cascadable{

    private Integer assessmentIndex = 0;
    private Boolean learningGoal = false;

    public SelfAssessment(String id) {
        super(id);
    }

    public SelfAssessment(String id, Integer assessmentIndex, Boolean learningGoal) {
        super(id);
        this.assessmentIndex = assessmentIndex;
        this.learningGoal = learningGoal;
    }

    public SelfAssessment(Integer assessmentIndex, Boolean learningGoal, User user, Competence competence) {
        super(user.getId() + competence.getId());
        this.assessmentIndex = assessmentIndex;
        this.learningGoal = learningGoal;
        this.user = user;
        this.competence = competence;
    }

    @Override
    public void persistMore() throws Exception {
        this.persist();
        createEdgeWith(CompObjectProperties.AssessmentOfCompetence, competence);
        createEdgeWith(CompObjectProperties.AssessmentOfUser, user);
    }

    public Integer getAssessmentIndex() {
        return assessmentIndex;
    }

    public Boolean getLearningGoal() {
        return learningGoal;
    }

    public void setAssessmentIndex(Integer assessmentIndex) {
        this.assessmentIndex = assessmentIndex;
    }

    public void setLearningGoal(Boolean learningGoal) {
        this.learningGoal = learningGoal;
    }

    public void addCompetenceToAssessment(Competence competence) throws Exception {
        createEdgeWith(CompObjectProperties.AssessmentOfCompetence, competence);
    }

    public void addUserToAssessment(User user) throws Exception {
        createEdgeWith(CompObjectProperties.AssessmentOfUser, user);
    }
}
